### PR TITLE
Adding DataLoader2 class as future replacement of DataLoader

### PR DESCRIPTION
**Pull request by VitalyFedyunin**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#63523 Adding DataLoader2 class as future replacement of DataLoader**
* #63522 Adding IterableAsDataPipe IterDataPipe
Supports sharding and batching on loader level**
* #63522 Adding IterableAsDataPipe IterDataPipe
usefull for tests and simple cases

Supports sharding and batching on loader level

Differential Revision: [D30426527](https://our.internmc.facebook.com/intern/diff/D30426527)

### CODE REVIEW

**Comment by VitalyFedyunin** [comment]
WIP: Need tests

File: torch/utils/data/dataloader_experimental.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+                datapipe = datapipe.shuffle()
+            if batch_outside_worker and pin_memory:
+                raise Exception(
+                    'pin_memory is not yet compatible with batch_outside_worker')
+            if not batch_outside_worker:
+                if batch_size is not None:
+                    datapipe = datapipe.batch(batch_size, drop_last=drop_last)
+                    if collate_fn is None:
+                        collate_fn = torch.utils.data._utils.collate.default_collate
+
```
Comment (ejguan) [inline torch/utils/data/dataloader_experimental.py:43]: Do we need to add a warning when the pipeline doesn't have `is_shardable` DataPipe with multiple workers?

**Review by ejguan — APPROVED** [review]
LGTM with one comment.

**Comment by VitalyFedyunin** [comment]
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30426527).

**Pull request closed without merging**